rpcs3/Emu/Io/Dimensions.cpp

```diff
@@ -0,0 +1,620 @@
+	// Confirmation is the first 4 bytes of the decrypted payload
+	const u32 conf = read_from_ptr<be_t<u32>>(value.data());
+	// Generate next random number based on RNG
+	const u32 next_random = get_next();
+	std::array<u8, 8> value_to_encrypt = {};
+	// Encrypt an 8 byte array, first 4 bytes are the next random number (little endian)
+	// followed by the confirmation from the decrypted payload
+	write_to_ptr<le_t<u32>>(value_to_encrypt.data(), next_random);
+	write_to_ptr<be_t<u32>>(value_to_encrypt.data(), 4, conf);
+	const std::array<u8, 8> encrypted = encrypt(value_to_encrypt.data(), std::nullopt);
+	reply_buf[0] = 0x55;
+	reply_buf[1] = 0x09;
+	reply_buf[2] = sequence;
+	// Copy encrypted value to response data
+	std::memcpy(&reply_buf[3], encrypted.data(), encrypted.size());
+	reply_buf[11] = generate_checksum(reply_buf, 11);
+}
+
+void dimensions_toypad::query_block(u8 index, u8 page, std::array<u8, 32>& reply_buf, u8 sequence)
+{
+	std::lock_guard lock(dimensions_mutex);
+
+	// Index from game begins at 1 rather than 0, so minus 1 here
+	dimensions_figure& figure = get_figure_by_index(index - 1);
+
+	reply_buf[0] = 0x55;
+	reply_buf[1] = 0x12;
+	reply_buf[2] = sequence;
+	reply_buf[3] = 0x00;
+	// Query 4 pages of 4 bytes from the figure, copy this to the response
+	if (figure.index != 255 && (4 * page) < ((0x2D * 4) - 16))
+	{
+		std::memcpy(&reply_buf[4], figure.data.data() + (4 * page), 16);
+	}
+	reply_buf[20] = generate_checksum(reply_buf, 20);
+}
+
+void dimensions_toypad::write_block(u8 index, u8 page, const u8* to_write_buf, std::array<u8, 32>& reply_buf, u8 sequence)
+{
+	std::lock_guard lock(dimensions_mutex);
+
+	// Index from game begins at 1 rather than 0, so minus 1 here
+	dimensions_figure& figure = get_figure_by_index(index - 1);
+
+	reply_buf[0] = 0x55;
+	reply_buf[1] = 0x02;
+	reply_buf[2] = sequence;
+	reply_buf[3] = 0x00;
+	// Copy 4 bytes to the page on the figure requested by the game
+	if (figure.index != 255 && page < 0x2D)
+	{
+		// Id is written to page 36
+		if(page == 36)
+		{
+			figure.id = read_from_ptr<le_t<u32>>(to_write_buf);
+		}
+		std::memcpy(figure.data.data() + (page * 4), to_write_buf, 4);
+		figure.save();
+	}
+	reply_buf[4] = generate_checksum(reply_buf, 4);
+}
+
+void dimensions_toypad::get_model(const u8* buf, u8 sequence, std::array<u8, 32>& reply_buf)
+{
+	// Decrypt payload to 8 byte array, byte 1 is the index, 4-7 are the confirmation
+	const std::array<u8, 8> value = decrypt(buf, std::nullopt);
+	const u8 index = value[0];
+	const u32 conf = read_from_ptr<be_t<u32>>(value.data(), 4);
+	// Index from game begins at 1 rather than 0, so minus 1 here
+	dimensions_figure& figure = get_figure_by_index(index - 1);
+	std::array<u8, 8> value_to_encrypt = {};
+	// Response is the figure's id (little endian) followed by the confirmation from payload
+	write_to_ptr<le_t<u32>>(value_to_encrypt.data(), figure.id);
+	write_to_ptr<be_t<u32>>(value_to_encrypt.data(), 4, conf);
```

Comment:
```suggestion
	const u32 conf = read_from_ptr<be_t<u32>>(value, 4);
	// Index from game begins at 1 rather than 0, so minus 1 here
	dimensions_figure& figure = get_figure_by_index(index - 1);
	std::array<u8, 8> value_to_encrypt = {};
	// Response is the figure's id (little endian) followed by the confirmation from payload
	write_to_ptr<le_t<u32>>(value_to_encrypt, figure.id);
	write_to_ptr<be_t<u32>>(value_to_encrypt, 4, conf);
```